specification/dartLangSpec.tex

```diff
@@ -986,46 +1047,69 @@ \section{Errors and Warnings}
-
-In a development environment a compiler should of course report compilation errors eagerly so as to best serve the programmer.
-
-A Dart development environment might choose to support error eliminating program transformations, e.g.,
+\rationale{%
+A Dart implementation has considerable freedom
+as to when compilation takes place.
+Modern programming language implementations
+often interleave compilation and execution,
+so that compilation of a method may be delayed, e.g.,
+until it is first invoked.
+Consequently, compile-time errors in a method $m$ may be reported
+as late as the time of $m$'s first invocation.
+
+Dart is often loaded directly from source,
+with no intermediate binary representation.
+In the interests of rapid loading, Dart implementations
+may choose to avoid full parsing of method bodies, for example.
+This can be done by tokenizing the input and
+checking for balanced curly braces on method body entry.
+In such an implementation, even syntax errors will be detected
+only when the method needs to be executed,
+at which time it will be compiled (JITed).
+
+In a development environment a compiler should of course
+report compilation errors eagerly so as to best serve the programmer.
+
+A Dart development environment might choose to support
+ error eliminating program transformations, e.g.,
 replacing an erroneous expression by the invocation of a debugger.
-It is outside the scope of this document to specify how such transformations work, and where they may be applied.
+It is outside the scope of this document
+to specify how such transformations work,
+and where they may be applied.%
 }
 
 \LMHash{}%
-If an uncaught compile-time error occurs within the code of a running isolate $A$, $A$ is immediately suspended.
-The only circumstance where a compile-time error could be caught would be via code run reflectively, where the mirror system can catch it.
+If an uncaught compile-time error occurs
+within the code of a running isolate $A$, $A$ is immediately suspended.
+The only circumstance where a compile-time error could be caught would be
+via code run reflectively, where the mirror system can catch it.
 
-\rationale{
-Typically, once a compile-time error is thrown and $A$ is suspended, $A$ will then be terminated.
+\rationale{%
+Typically, once a compile-time error is thrown and $A$ is suspended,
+$A$ will then be terminated.
 However, this depends on the overall environment.
 A Dart engine runs in the context of an \Index{embedder},
-a program that interfaces between the engine and the surrounding computing environment.
-The embedder will often be a web browser, but need not be; it may be a C++ program on the server for example.
-When an isolate fails with a compile-time error as described above, control returns to the embedder, along with an exception describing the problem.
+a program that interfaces between the engine and
+the surrounding computing environment.
+The embedder will often be a web browser, but need not be;
```

Comment:
Phew, good thing it doesn't need to be one, since it *never is*.
(Might be worth fixing this sentence 😁)

Reply:
Adjusted this sentence to drop the web browser part, and changed the word 'embedder' to 'runtime'.

You could say that 'runtime' is more ambiguous than 'runtime system', but I think we'll be OK because it is defined explicitly at this location, and it can be looked up in the index. WDYT?

Reply:
LGTM

Reply:
Thanks!